web/packages/shared/components/TerminalSearch/TerminalSearch.tsx

```diff
@@ -0,0 +1,180 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { useEffect, useState, useRef, useCallback } from 'react';
+import styled, { useTheme } from 'styled-components';
+import { Flex, ButtonIcon, Box, P2, Input } from 'design';
+import { ArrowDown, ArrowUp, Cross } from 'design/Icon';
+import { SearchAddon } from '@xterm/addon-search';
+
+export interface TerminalWithSearch {
+  getSearchAddon({
+    onDidChangeResults,
+  }: {
+    onDidChangeResults: ({
+      resultIndex,
+      resultCount,
+    }: {
+      resultIndex: number;
+      resultCount: number;
+    }) => void;
+  }): SearchAddon;
+}
+
+export const TerminalSearch = ({
+  xTerm,
+  show,
+  onClose,
+  onOpen,
+  focusTerm,
+  isSearchKeyboardEvent,
+}: {
+  xTerm: TerminalWithSearch;
+  show: boolean;
+  focusTerm(): void;
+  onClose(): void;
+  onOpen(): void;
+  isSearchKeyboardEvent: (e: KeyboardEvent) => boolean;
+}) => {
+  const theme = useTheme();
+  const searchInputRef = useRef<HTMLInputElement>();
+  const [searchValue, setSearchValue] = useState('');
+  const [searchResults, setSearchResults] = useState<{
+    resultIndex: number;
+    resultCount: number;
+  }>({ resultIndex: 0, resultCount: 0 });
+
+  const searchAddon = xTerm?.getSearchAddon({
+    onDidChangeResults: setSearchResults,
+  });
+
+  const search = (value: string, direction: 'next' | 'previous') => {
+    const match = theme.colors.terminal.brightYellow;
+    const activeMatch = theme.colors.terminal.yellow;
+    setSearchValue(value);
+    const opts = {
+      regex: true,
+      caseSensitive: false,
+      decorations: {
+        matchOverviewRuler: match,
+        activeMatchColorOverviewRuler: activeMatch,
+        matchBackground: match,
+        activeMatchBackground: activeMatch,
+      },
+    };
+
+    if (direction === 'next') {
+      searchAddon.findNext(value, opts);
+    } else {
+      searchAddon.findPrevious(value, opts);
+    }
+  };
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    search(e.target.value, 'next');
+  };
+
+  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      search(e.currentTarget.value, 'next');
+    }
+  };
+
+  const searchNext = () => {
+    search(searchValue, 'next');
+  };
+
+  const searchPrevious = () => {
+    search(searchValue, 'previous');
+  };
+
+  const onEscape = useCallback(() => {
+    onClose();
+    searchAddon.clearDecorations();
+    focusTerm();
+  }, [onClose, searchAddon, focusTerm]);
+
+  useEffect(() => {
+    const listener = (e: KeyboardEvent) => {
+      if (e.key === 'Escape') {
+        onEscape();
+        e.preventDefault();
+      }
+      if (isSearchKeyboardEvent(e)) {
+        onOpen();
+        searchInputRef.current?.focus();
+        e.preventDefault();
+      }
+    };
+
+    document.addEventListener('keydown', listener);
```

Comment:
This may look a little different. I haven't been able to quite figure it out yet, but we can't pass the `attachCustomKeyEventHandler` to this addon because then it will overwrite any previous callbacks that were set (like the `copy/paste` one set for Connect).

So we'll have to do something where this keyboard event handler calls maybe some map of function=>boolean and iterates through that instead.

or i'll add some sort of callbacks that can be hooked into in `terminal.ts`/`ctrl.ts`

Reply:
ok,. fixed it. it now works 'per terminal' in web and connect

Reply:
Ah, I've made a mistake. We can't call `attachCustomKeyEventHandler` directly because it overwrites the existing handler (I think you commented about it, but I can't find it). We shouldn't also expose the internal `term` directly to `TerminalSearch`. Instead we should have a list of event handlers (same for Web UI):

```
diff --git a/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx b/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx
index 08a259bd44..630c07cde8 100644
--- a/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx
+++ b/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx
@@ -24,11 +24,9 @@ import { SearchAddon } from '@xterm/addon-search';
 
 export interface TerminalSeacher {
   getSearchAddon(): SearchAddon;
-  term: {
-    focus(): void;
-    attachCustomKeyEventHandler: (
-      customEvent: (e: KeyboardEvent) => boolean
-    ) => void;
+  focus(): void;
+  registerCustomKeyEventHandler(customEvent: (e: KeyboardEvent) => boolean): {
+    unregister(): void;
   };
 }
 
@@ -106,11 +104,11 @@ export const TerminalSearch = ({
   const onEscape = useCallback(() => {
     onClose();
     xTerm.getSearchAddon().clearDecorations();
-    xTerm.term.focus();
+    xTerm.focus();
   }, [onClose, xTerm]);
 
   useEffect(() => {
-    xTerm.term.attachCustomKeyEventHandler(e => {
+    const { unregister } = xTerm.registerCustomKeyEventHandler(e => {
       if (isSearchKeyboardEvent(e)) {
         onOpen();
         searchInputRef.current?.focus();
@@ -126,6 +124,8 @@ export const TerminalSearch = ({
 
       return true;
     });
+
+    return unregister;
   }, [onEscape, onOpen, searchInputRef, xTerm, isSearchKeyboardEvent]);
 
   if (!show) {
diff --git a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts
index 54bea422c2..ab763d7708 100644
--- a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts
+++ b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts
@@ -27,6 +27,7 @@ import { IPtyProcess } from 'teleterm/sharedProcess/ptyHost';
 import Logger from 'teleterm/logger';
 import { ConfigService, AppConfig } from 'teleterm/services/config';
 import { KeyboardShortcutsService } from 'teleterm/ui/services/keyboardShortcuts';
+import { TerminalSeacher } from 'shared/components/TerminalSearch/TerminalSearch';
 
 const WINDOW_RESIZE_DEBOUNCE_DELAY = 200;
 
@@ -38,7 +39,7 @@ type Options = {
   openContextMenu(e: MouseEvent): void;
 };
 
-export default class TtyTerminal {
+export default class TtyTerminal implements TerminalSeacher {
   public term: Terminal;
   private el: HTMLElement;
   private fitAddon = new FitAddon();
@@ -51,6 +52,8 @@ export default class TtyTerminal {
     AppConfig,
     'terminal.rightClick' | 'terminal.copyOnSelect'
   >;
+  private customKeyEventHandlers: Set<(event: KeyboardEvent) => boolean> =
+    new Set();
 
   constructor(
     private ptyProcess: IPtyProcess,
@@ -71,6 +74,13 @@ export default class TtyTerminal {
     );
   }
 
+  registerCustomKeyEventHandler(customHandler: (e: KeyboardEvent) => boolean) {
+    this.customKeyEventHandlers.add(customHandler);
+    return {
+      unregister: () => this.customKeyEventHandlers.delete(customHandler),
+    };
+  }
+
   open(): void {
     this.term = new Terminal({
       cursorBlink: false,
@@ -109,7 +119,7 @@ export default class TtyTerminal {
 
     this.term.open(this.el);
 
-    this.term.attachCustomKeyEventHandler(e => {
+    this.registerCustomKeyEventHandler(e => {
       const action = this.keyboardShortcutsService.getShortcutAction(e);
       const isKeyDown = e.type === 'keydown';
       if (action === 'terminalCopy' && isKeyDown && this.term.hasSelection()) {
@@ -130,6 +140,17 @@ export default class TtyTerminal {
       return true;
     });
 
+    this.term.attachCustomKeyEventHandler(e => {
+      for (const eventHandler of this.customKeyEventHandlers) {
+        if (!eventHandler(e)) {
+          // The event was handled, we can return early.
+          return false;
+        }
+      }
+      // The event wasn't handled, pass it to xterm.
+      return true;
+    });
+
     this.term.element.addEventListener('contextmenu', e => {
       // We always call preventDefault because:
       // 1. When `terminalRightClick` is not `menu`, we don't want to show it.

```